### PR TITLE
Support for extra postcss plugins

### PR DESCRIPTION
**Pull request by mvsmal**
Added support for extra postcss plugins. They can be specified in `"filetypes"` section of config like this:
```json
"filetypes": {
  ".scss": {
    "syntax": "postcss-scss",
    "plugins": ["postcss-nested"]
  }
}

### CODE REVIEW

File: src/requireCssModule.js [hunk 1]
```diff
@@ -18,19 +18,57 @@ import type {
   StyleModuleMapType
 } from './types';
 
-const getTokens = (runner, cssSourceFilePath: string, filetypes): StyleModuleMapType => {
+type FileTypeOptions = {|
+  syntax: string,
+  plugins: Array<string>
```
Comment (gajus) [inline src/requireCssModule.js]: Please use `$ReadOnlyArray` and covariant props, i.e.  `{|+syntax: string, plugins: $ReadOnlyArray<string>|}`

File: src/requireCssModule.js [hunk 2]
```diff
@@ -18,19 +18,57 @@ import type {
...
+|};
+
+const getFiletypeOptions = (cssSourceFilePath: string, filetypes: Object): ?(string|FileTypeOptions) => {
```
Comment (gajus) [inline src/requireCssModule.js]: `filetypes` type needs to be `{[key: string]: FileTypeOptions}`

Comment (gajus) [inline src/requireCssModule.js]: `?(string|FileTypeOptions)` does not match the return type (not getting an error here?)  `FileTypeOptions | string | null`.

Where does the `string` come from, though?

Comment (gajus) [inline src/requireCssModule.js]: > Where does the string come from, though?

I see.

Lets remove the shortcut. This is going to be a breaking change.

Require that `FileTypeOptions` is always present.

File: src/requireCssModule.js [hunk 3]
```diff
@@ -18,19 +18,57 @@ import type {
...
-  const syntax = filetypes[extension];
+  const filetype = filetypes ? filetypes[extension] : null;
 
-  const options: Object = {
-    from: cssSourceFilePath
-  };
+  return filetype;
+};
+
+const getSyntax = (filetypeOptions: ?(string|FileTypeOptions)) => {
```
Comment (gajus) [inline src/requireCssModule.js]: Require that `filetypeOptions` is always `FileTypeOptions`. Do not call `getSyntax` if `filetypeOptions` option is void.

File: src/requireCssModule.js [hunk 4]
```diff
@@ -18,19 +18,57 @@ import type {
...
+  if (typeof filetypeOptions === 'object') {
     // eslint-disable-next-line import/no-dynamic-require, global-require
-    options.syntax = require(syntax);
+    return require(filetypeOptions.syntax);
+  }
+
+  return null;
+};
+
+const getExtraPlugins = (filetypeOptions: ?(string|FileTypeOptions)): Array<any> => {
```
Comment (gajus) [inline src/requireCssModule.js]: Use `$ReadOnlyArray`

File: src/requireCssModule.js (hunk 3 above)
Comment (gajus) [inline src/requireCssModule.js]: Declare the return type annotation for `getSyntax`.

File: test/fixtures/react-css-modules/resolves less stylesheets/options.json [hunk 5]
```diff
@@ -5,7 +5,9 @@
       {
         "generateScopedName": "[name]__[local]",
         "filetypes": {
-          ".less": "postcss-less"
+          ".less": {
+            "syntax": "postcss-less"
```
Comment (gajus) [inline test/fixtures/react-css-modules/resolves less stylesheets/options.json]: This is using the old syntax.

Comment (mvsmal) [inline test/fixtures/react-css-modules/resolves less stylesheets/options.json]: I've changed it to Object `{ "syntax": "..." }`. Sorry, what's wrong?

Comment (gajus) [inline test/fixtures/react-css-modules/resolves less stylesheets/options.json]: Oops. Sorry. This PR was the first thing I looked at this morning. :-)

**Comment by mvsmal** [comment]
So. What's next? I've updated the code according to your requests.

**Comment by gajus** [comment]
> So. What's next? I've updated the code according to your requests.

Sorry, overloaded at work. Merged.

Thank you,

**Comment by Slapbox** [comment]
Is there any way to set options on the plugins loaded this way?

**Pull request merged**